integration/rootless: add more asserts

wainersm opened the pull request:
This added two more asserts to the rootless test:
- that the QEMU process is owned by the temporary non-root user
- that the QEMU files are owned by that user too.

Fixes #4654
Signed-off-by: Wainer dos Santos Moschetta <wainersm@redhat.com>

devimc commented:
thanks @wainersm but afaik the CI job where this test runs was removed

wainersm commented:
Hi @devimc ,

> thanks @wainersm but afaik the CI job where this test runs was removed

It's executed in kata-containers-fedora-35-x86_64-CRIO_K8S-PR at least. For example, http://jenkins.katacontainers.io/view/PR%20Triggered/job/kata-containers-fedora-35-x86_64-CRIO_K8S-PR/341/console : 
```
08:16:37 INFO: Running rootless tests
08:16:37 bash -f integration/rootless/rootless_test.sh
08:16:37 "SANDBOX_READY"
08:16:39 Stopped sandbox 93a477025da9f9c1ff3c3a605336457b16b20db0568808429a6d39e62b5bda4b
08:16:39 Removed sandbox 93a477025da9f9c1ff3c3a605336457b16b20db0568808429a6d39e62b5bda4b
```

Am I missing something?

wainersm commented:
/test-fedora

devimc reviewed (approved):
thanks @wainersm

snir911 commented on integration/rootless/rootless_test.sh:
```diff
@@ -27,19 +27,35 @@ setup() {
+	[[ -n "$kata_uid" && -n "$kata_gid" ]]
+
+	# Assert the QEMU process is owned by that user.
+	ps -u "$kata_uid" | grep "qemu" &>/dev/null || \
```
should we validate qemu process exists first?

wainersm replied:
I'm not sure. And if there is a running qemu process, it is hard to tell for sure it belongs to this test case. However your comment made me review the code and I found some flaws....gonna send a v2.

wainersm commented:
Update this PR with:
 * added a check for the case there are more than one kata UID

dborquez reviewed (approved):
LGTM, thanks @wainersm. I saw an improvement but then you fixed it.

wainersm commented:
> LGTM, thanks @wainersm. I saw an improvement but then you fixed it.

hey @dborquez , thanks for reviewing!

wainersm commented:
/test

snir911 reviewed (approved):
thanks!

gkurz commented on integration/rootless/rootless_test.sh:
```diff
@@ -27,19 +27,39 @@ setup() {
+		die "Unabled to find the kata temporary UID"
+
+	# Assert the QEMU process is owned by that user.
+	ps -u "$kata_uid" | grep "qemu" &>/dev/null || \
```
Is it possible that some other unrelated process gets caught by this `grep "qemu"`  ? What about extracting the name of the QEMU executable out of `configuration.toml` and passing its `basename` to something like `ps -C ${qemu_binary} -o uid --no-headers` ?

c3d replied:
Isn't that the case that if there is more than one process, it's likely to be another qemu instance? How would your ps help? IOW, what are you trying to filter out? A "qemu-img" process for example?

c3d replied:
Ah, unless you were thinking about the frequent case where a `ps | grep` finds the `grep` process. But then that cannot happen here because the `grep` does not run as user `$kata_uid`.

wainersm replied:
> Is it possible that some other unrelated process gets caught by this `grep "qemu"` ? What about extracting the name of the QEMU executable out of `configuration.toml` and passing its `basename` to something like `ps -C ${qemu_binary} -o uid --no-headers` ?

The way I understood rootless, kata is going to create a temporary user (kata-`RANDOM_ID`) for every and each sandbox. The QEMU process and some control files should be owned by that user. Here I list all the processes owned by the that temporary user, and checking there is a match to "qemu". It is hard to tell exactly what is the binary name because the build scripts allow users to overwrite it; for example, there are jobs using "qemu-tdx". An alternative is to get the path to the hypervisor set on kata's config file.

Do you think it should be more accurate?

wainersm replied:
ah, there is also another approach: call the `check_processes()` on setup and fail if any QEMU process is actually running at test time. It should work fine with CI but it mightl make developers unhappy when they are running the tests locally.

gkurz replied:
> > Is it possible that some other unrelated process gets caught by this `grep "qemu"` ? What about extracting the name of the QEMU executable out of `configuration.toml` and passing its `basename` to something like `ps -C ${qemu_binary} -o uid --no-headers` ?
> 
> The way I understood rootless, kata is going to create a temporary user (kata-`RANDOM_ID`) for every and each sandbox. The QEMU process and some control files should be owned by that user. Here I list all the processes owned by the that temporary user, and checking there is a match to "qemu". It is hard to tell exactly what is the binary name because the build scripts allow users to overwrite it; for example, there are jobs using "qemu-tdx". An alternative is to get the path to the hypervisor set on kata's config file.

My suggestion is precisely to get the path to QEMU from the config file :smiley:

> 
> Do you think it should be more accurate?

Yes.

wainersm replied:
@gkurz ok, what do you think now?

wainersm commented:
Updated this PR with:
 - call `extract_kata_env` on `setup` then use `HYPERVISOR_PATH` to grep the process with better accuracy (address @gkurz 's suggestion)

gkurz commented:
/test

gkurz commented on integration/rootless/rootless_test.sh:
```diff
@@ -27,19 +28,40 @@ setup() {
+	local kata_uid="$(eval $grep_uid | cut -d: -f3)"
+	local kata_gid="$(eval $grep_uid | cut -d: -f4)"
+	[[ -n "$kata_uid" && -n "$kata_gid" ]] || \
+		die "Unabled to find the kata temporary UID"
```
s/Unabled/Unable

gkurz commented on integration/rootless/rootless_test.sh:
```diff
@@ -27,19 +28,40 @@ setup() {
+
+	# Assert the QEMU process is owned by that user.
+	ps -u "$kata_uid" -o cmd --no-header | \
+		grep -q "^${HYPERVISOR_PATH:-qemu}" || \
```
If the QEMU path isn't present in the config file, `kata-runtime kata-env` falls back on the hard-coded default value `qemu-system-XXXX`  => `HYPERVISOR_PATH` is always set and non-null AFAICT.

wainersm replied:
I was worried about a bug on `extract_kata_env` that would export a empty string.

gkurz commented on integration/rootless/rootless_test.sh:
```diff
@@ -27,19 +28,40 @@ setup() {
+		grep -q "^${HYPERVISOR_PATH:-qemu}" || \
+		die "Expects the QEMU process be owned by a non-root user"
+
+	# Assert the QEMU files are owned by that user too.
```
How are we sure that QEMU has created any file at this point ?

wainersm replied:
you are right. Gonna fix it.

wainersm commented:
Updated this PR:
 - To address the comment of @gkurz about the necessity to check if the QEMU files really exist.

gkurz commented on integration/rootless/rootless_test.sh:
```diff
@@ -27,19 +28,45 @@ setup() {
+		die "Unable to find the pod directory: $pod_dir"
+	local pod_files=($(sudo find "$pod_dir"))
+	[ "${#pod_files[@]}" -gt 0 ] || \
+		die "No pod files found in the directory: $pod_dir"
```
Shouldn't we `waitForProcess` for the pod dir and files to show up ?

wainersm replied:
At this point the files should already exist otherwise something is wrong because on line 37 we wait the pod to show up (https://github.com/kata-containers/tests/pull/4655/commits/b1716d9b5d2823984797110946782708a7a9d586#diff-c969ae5cb95b0b60400857625d05b7e33264a94d3fadbef39ca9db6c151d299fR37)

gkurz replied:
@wainersm are you saying that a POD reaching SANDBOX_READY guarantees that the QEMU process already created its pid file, QMP socket ... etc... ?

gkurz replied:
> @wainersm are you saying that a POD reaching SANDBOX_READY guarantees that the QEMU process already created its pid file, QMP socket ... etc... ?

@jodh-intel @egernst @fidencio any insights to share ?

fidencio replied:
Take this with a grain of salt, okay?

My understanding is that when we reach SANDBOX_READY we're good to go, it means the SANDBOX was created but the containers inside it are not yet started, but from the sandbox perspective (VMM and its helper processes) we're fine.

fidencio replied:
Just to complement: https://github.com/kata-containers/kata-containers/blob/604a7950738ac7b6d2d621fa262f6ada7f718aef/src/runtime/virtcontainers/sandbox.go#L474-L516 is where we set the StateReady for the sandbox.

Earlier in that function, we call: https://github.com/kata-containers/kata-containers/blob/604a7950738ac7b6d2d621fa262f6ada7f718aef/src/runtime/virtcontainers/sandbox.go#L518-L607, which is responsible for creating the VM (see: https://github.com/kata-containers/kata-containers/blob/604a7950738ac7b6d2d621fa262f6ada7f718aef/src/runtime/virtcontainers/sandbox.go#L597-L600)

So, yeah, we're good to go.

gkurz replied:
> So, yeah, we're good to go.

I'm not that sure because `createSanbox()` is [called **before** QEMU is even started](https://github.com/kata-containers/kata-containers/blob/604a7950738ac7b6d2d621fa262f6ada7f718aef/src/runtime/virtcontainers/api.go#L61-L91) by `startVM()`. If the `StateReady` state is visible in the meantime then we cannot assume that QEMU has done anything at all.

fidencio replied:
You're right, @gkurz.

We should be looking for `StateRunning` in the Sandbox, as `s.hypervisor.CreateVM()` will build the QEMU command line and **not** start the VM, my bad here.

wainersm replied:
Thanks @gkurz and @fidencio ! this discussion has been very educational to me. I will change the script to wait on the files to show up with a small timeout.

wainersm commented:
Updated this PR with:
 * Added  loop through known files (console.sock qmp.sock pid vhost-fs.sock) and `waitForProcess` to wait the file be created before checking its ownership

@gkurz Does it address your concerns?

gkurz commented on integration/rootless/rootless_test.sh:
```diff
@@ -27,19 +28,46 @@ setup() {
+		waitForProcess 10 1 "test -e "$file"" || \
+			die "File didn't show up: $file"
+		[ "$(sudo stat -c %u-%g $file)" == "${kata_uid}-${kata_gid}" ]
+	done
```
> Updated this PR with:
> 
>     * Added  loop through known files (console.sock qmp.sock pid vhost-fs.sock) and `waitForProcess` to wait the file be created before checking its ownership
> 
> 
> @gkurz Does it address your concerns?

Yes for the files but according to our findings, it is possible that the QEMU process hasn't even started yet. It seems that the `ps` check should also go through `waitForProcess`.

wainersm replied:
oh, True. Forgot that.

wainersm commented:
Updated this PR with:
 * added a couple of `waitForProcess` to ensure resources are available

wainersm commented:
/test

gkurz reviewed (approved):
This looks fine and race-free now :smiley:. Thanks @wainersm !

wainersm commented:
> This looks fine and race-free now smiley. Thanks @wainersm !

Thanks for the very good reviews, @gkurz !

wainersm commented:
/test-devmapper

fidencio commented:
/retest-devmapper-clh

gkurz commented:
/test-devmapper

fidencio commented:
/retest-devmapper

gkurz commented:
/retest-devmapper-clh
